File: apps/demo-go/standardized_test.go
package main

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/parallax/sdk-go/pkg/parallax"
)

// TestAgent implements the standardized test agent
type TestAgent struct {
	parallax.BaseAgent
}

// NewTestAgent creates a new test agent
func NewTestAgent() *TestAgent {
	return &TestAgent{
		BaseAgent: parallax.BaseAgent{
			ID:           "test-agent-go",
			Name:         "Test Agent (Go)",
			Capabilities: []string{"analysis", "validation"},
			Expertise:    0.85,
		},
	}
}

// Analyze implements the agent analyze method
func (a *TestAgent) Analyze(ctx context.Context, task string, input map[string]interface{}) (*parallax.AgentResponse, error) {
	switch task {
	case "analyze":
		data, _ := input["data"].(map[string]interface{})
		contentType := "unknown"
		if t, ok := data["type"].(string); ok {
			contentType = t
		}
		content := ""
		if c, ok := data["content"].(string); ok {
			content = c
		}

		return &parallax.AgentResponse{
			Value: map[string]interface{}{
				"summary": fmt.Sprintf("Analyzed %s content", contentType),
				"length":  len(content),
				"result":  "Analysis complete",
			},
			Confidence: 0.85,
			Reasoning:  "Standard analysis performed",
		}, nil

	case "validate":
		data, _ := input["data"].(map[string]interface{})
		value, _ := data["value"].(float64) // JSON numbers are float64
		rules, _ := data["rules"].([]interface{})
		
		details := []string{}
		valid := true

		for _, r := range rules {
			rule, _ := r.(string)
			switch rule {
			case "positive":
				if value > 0 {
					details = append(details, "Value is positive")
				} else {
					valid = false
					details = append(details, "Value is not positive")
				}
			case "even":
				if int(value)%2 == 0 {
					details = append(details, "Value is even")
				} else {
					valid = false
					details = append(details, "Value is not even")
				}
			}
		}

		return &parallax.AgentResponse{
			Value: map[string]interface{}{
				"valid":   valid,
				"details": details,
			},
			Confidence: 0.95,
			Reasoning:  "Validation rules applied",
		}, nil

	default:
		return nil, fmt.Errorf("unknown task: %s", task)
	}
}

func runStandardizedTests() bool {
	fmt.Println("=== Parallax SDK Test Results ===")
	fmt.Println("Language: Go")
	fmt.Println("SDK Version: 0.1.0\n")

	results := make(map[string]bool)

	// Test 1: Agent Creation
	func() {
		defer func() {
			if r := recover(); r != nil {
				results["Agent Creation"] = false
				fmt.Printf("Test 1: Agent Creation............... FAIL (%v)\n", r)
			}
		}()

		agent := NewTestAgent()
		passed := agent.ID == "test-agent-go" &&
			contains(agent.Capabilities, "analysis") &&
			contains(agent.Capabilities, "validation")
		results["Agent Creation"] = passed
		status := "PASS"
		if !passed {
			status = "FAIL"
		}
		fmt.Printf("Test 1: Agent Creation............... %s\n", status)
	}()

	// Test 2: Simple Analysis
	func() {
		defer func() {
			if r := recover(); r != nil {
				results["Simple Analysis"] = false
				fmt.Printf("Test 2: Simple Analysis.............. FAIL (%v)\n", r)
			}
		}()

		agent := NewTestAgent()
		response, err := agent.Analyze(context.Background(), "analyze", map[string]interface{}{
			"data": map[string]interface{}{
				"content": "Test data for analysis",
				"type":    "text",
			},
		})
		
		passed := err == nil && response.Confidence >= 0.7 && response.Value != nil
		results["Simple Analysis"] = passed
		status := "PASS"
		if !passed {
			status = "FAIL"
		}
		fmt.Printf("Test 2: Simple Analysis.............. %s\n", status)
	}()

	// Test 3: Validation
	func() {
		defer func() {
			if r := recover(); r != nil {
				results["Validation"] = false
				fmt.Printf("Test 3: Validation................... FAIL (%v)\n", r)
			}
		}()

		agent := NewTestAgent()
		response, err := agent.Analyze(context.Background(), "validate", map[string]interface{}{
			"data": map[string]interface{}{
				"value": 42.0,
				"rules": []interface{}{"positive", "even"},
			},
		})
		
		value := response.Value.(map[string]interface{})
		details := value["details"].([]string)
		passed := err == nil &&
			value["valid"].(bool) == true &&
			response.Confidence == 0.95 &&
			len(details) == 2
		results["Validation"] = passed
		status := "PASS"
		if !passed {
			status = "FAIL"
		}
		fmt.Printf("Test 3: Validation................... %s\n", status)
	}()

	// Test 4: Error Handling
	func() {
		defer func() {
			if r := recover(); r != nil {
				results["Error Handling"] = false
				fmt.Printf("Test 4: Error Handling............... FAIL (%v)\n", r)
			}
		}()

		agent := NewTestAgent()
		_, err := agent.Analyze(context.Background(), "unknown-task", map[string]interface{}{})
		
		if err == nil {
			results["Error Handling"] = false
			fmt.Println("Test 4: Error Handling............... FAIL (No error thrown)")
		} else {
			passed := strings.Contains(strings.ToLower(err.Error()), "unknown task")
			results["Error Handling"] = passed
			status := "PASS"
			if !passed {
				status = "FAIL"
			}
			fmt.Printf("Test 4: Error Handling............... %s\n", status)
		}
	}()

	// Test 5: Client API (optional)
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Println("Test 5: Client API (optional)........ SKIP (Control plane not running)")
			}
		}()

		client, err := parallax.NewClient(parallax.ClientConfig{
			BaseURL: "http://localhost:8080",
		})
		if err != nil {
			fmt.Println("Test 5: Client API (optional)........ SKIP (Control plane not running)")
			return
		}

		ctx := context.Background()
		
		// 5.1 Health Check
		health, err := client.Health(ctx)
		if err != nil {
			fmt.Println("Test 5: Client API (optional)........ SKIP (Control plane not running)")
			return
		}

		// 5.2 List Patterns
		patterns, err := client.ListPatterns(ctx)
		if err != nil {
			results["Client API"] = false
			fmt.Println("Test 5: Client API (optional)........ FAIL")
			return
		}

		// 5.3 Pattern Execution
		execution, err := client.ExecutePattern(ctx, "SimpleConsensus", map[string]interface{}{
			"task": "SDK test",
			"data": map[string]interface{}{"test": true},
		})

		passed := health != nil && len(patterns) > 0 && execution.ID != ""
		results["Client API"] = passed
		status := "PASS"
		if !passed {
			status = "FAIL"
		}
		fmt.Printf("Test 5: Client API (optional)........ %s\n", status)
	}()

	// Summary
	passed := 0
	for _, v := range results {
		if v {
			passed++
		}
	}
	total := len(results)
	fmt.Printf("\nSummary: %d/%d tests passed\n", passed, total)

	return passed == total
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

func main() {
	success := runStandardizedTests()
	if !success {
		os.Exit(1)
	}
}

File: apps/demo-rust/src/standardized_test.rs
use anyhow::{anyhow, Result};
use serde_json::json;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::time::sleep;
use std::time::Duration;

// Import from the workspace package
use parallax_sdk::{ParallaxAgent, AgentResult};

/// Test agent implementation using ParallaxAgent
struct TestAgent {
    agent: Arc<ParallaxAgent>,
}

impl TestAgent {
    fn new() -> Self {
        let agent = ParallaxAgent::new(
            "test-agent-rust",
            "Test Agent (Rust)",
            vec!["analysis".to_string(), "validation".to_string()],
            {
                let mut metadata = HashMap::new();
                metadata.insert("expertise".to_string(), "0.85".to_string());
                metadata
            },
        ).set_analyze_fn(move |task: &str, data: Option<serde_json::Value>| -> futures::future::BoxFuture<'static, Result<AgentResult, Box<dyn std::error::Error>>> {
            let task = task.to_string();
            Box::pin(async move {
                match task.as_str() {
                    "analyze" => {
                        let data = data.ok_or("Missing data")?;
                        let data_obj = data.as_object().ok_or("Data is not an object")?;
                        let content_data = data_obj.get("data")
                            .and_then(|v| v.as_object())
                            .ok_or("Missing data field")?;
                        
                        let content_type = content_data.get("type")
                            .and_then(|v| v.as_str())
                            .unwrap_or("unknown");
                        let content = content_data.get("content")
                            .and_then(|v| v.as_str())
                            .unwrap_or("");

                        Ok(AgentResult {
                            value: json!({
                                "summary": format!("Analyzed {} content", content_type),
                                "length": content.len(),
                                "result": "Analysis complete"
                            }),
                            confidence: 0.85,
                            reasoning: Some("Standard analysis performed".to_string()),
                            uncertainties: vec![],
                            metadata: HashMap::new(),
                        })
                    }

                    "validate" => {
                        let data = data.ok_or("Missing data")?;
                        let data_obj = data.as_object().ok_or("Data is not an object")?;
                        let validate_data = data_obj.get("data")
                            .and_then(|v| v.as_object())
                            .ok_or("Missing data field")?;
                        
                        let value = validate_data.get("value")
                            .and_then(|v| v.as_i64())
                            .ok_or("Missing value")?;
                        let rules = validate_data.get("rules")
                            .and_then(|v| v.as_array())
                            .ok_or("Missing rules")?;

                        let mut details = Vec::new();
                        let mut valid = true;

                        for rule in rules {
                            if let Some(rule_str) = rule.as_str() {
                                match rule_str {
                                    "positive" => {
                                        if value > 0 {
                                            details.push("Value is positive");
                                        } else {
                                            valid = false;
                                            details.push("Value is not positive");
                                        }
                                    }
                                    "even" => {
                                        if value % 2 == 0 {
                                            details.push("Value is even");
                                        } else {
                                            valid = false;
                                            details.push("Value is not even");
                                        }
                                    }
                                    _ => {}
                                }
                            }
                        }

                        Ok(AgentResult {
                            value: json!({
                                "valid": valid,
                                "details": details
                            }),
                            confidence: 0.95,
                            reasoning: Some("Validation rules applied".to_string()),
                            uncertainties: vec![],
                            metadata: HashMap::new(),
                        })
                    }

                    _ => Err(format!("Unknown task: {}", task).into())
                }
            })
        });
        
        Self {
            agent: Arc::new(agent),
        }
    }
    
    async fn analyze(&self, task: &str, data: serde_json::Value) -> Result<AgentResult> {
        (self.agent.analyze_fn)(task, Some(data)).await.map_err(|e| anyhow!("{}", e))
    }
}

async fn run_standardized_tests() -> Result<bool> {
    println!("=== Parallax SDK Test Results ===");
    println!("Language: Rust");
    println!("SDK Version: 0.1.0\n");

    let mut results = HashMap::new();

    // Test 1: Agent Creation
    let test1_result = (|| -> Result<bool> {
        let test_agent = TestAgent::new();
        Ok(test_agent.agent.id == "test-agent-rust" &&
           test_agent.agent.capabilities.contains(&"analysis".to_string()) &&
           test_agent.agent.capabilities.contains(&"validation".to_string()))
    })();
    
    match test1_result {
        Ok(passed) => {
            results.insert("Agent Creation", passed);
            println!("Test 1: Agent Creation............... {}", if passed { "PASS" } else { "FAIL" });
        }
        Err(e) => {
            results.insert("Agent Creation", false);
            println!("Test 1: Agent Creation............... FAIL ({})", e);
        }
    }

    // Test 2: Simple Analysis
    let test2_result: Result<bool> = (|| async {
        let test_agent = TestAgent::new();
        let response = test_agent.analyze("analyze", json!({
            "data": {
                "content": "Test data for analysis",
                "type": "text"
            }
        })).await?;
        Ok(response.confidence >= 0.7 && !response.value.is_null())
    })().await;

    match test2_result {
        Ok(passed) => {
            results.insert("Simple Analysis", passed);
            println!("Test 2: Simple Analysis.............. {}", if passed { "PASS" } else { "FAIL" });
        }
        Err(e) => {
            results.insert("Simple Analysis", false);
            println!("Test 2: Simple Analysis.............. FAIL ({})", e);
        }
    }

    // Test 3: Validation
    let test3_result: Result<bool> = (|| async {
        let test_agent = TestAgent::new();
        let response = test_agent.analyze("validate", json!({
            "data": {
                "value": 42,
                "rules": ["positive", "even"]
            }
        })).await?;
        
        let valid = response.value["valid"].as_bool().unwrap_or(false);
        let details = response.value["details"].as_array().map(|a| a.len()).unwrap_or(0);
        
        Ok(valid && response.confidence == 0.95 && details == 2)
    })().await;

    match test3_result {
        Ok(passed) => {
            results.insert("Validation", passed);
            println!("Test 3: Validation................... {}", if passed { "PASS" } else { "FAIL" });
        }
        Err(e) => {
            results.insert("Validation", false);
            println!("Test 3: Validation................... FAIL ({})", e);
        }
    }

    // Test 4: Error Handling
    let test4_result: Result<bool> = (|| async {
        let test_agent = TestAgent::new();
        match test_agent.analyze("unknown-task", json!({})).await {
            Err(e) => Ok(e.to_string().to_lowercase().contains("unknown task")),
            Ok(_) => Ok(false),
        }
    })().await;

    match test4_result {
        Ok(passed) => {
            results.insert("Error Handling", passed);
            if !passed {
                println!("Test 4: Error Handling............... FAIL (No error thrown)");
            } else {
                println!("Test 4: Error Handling............... PASS");
            }
        }
        Err(e) => {
            results.insert("Error Handling", false);
            println!("Test 4: Error Handling............... FAIL ({})", e);
        }
    }

    // Test 5: gRPC Server and Registration
    print!("Test 5: gRPC Server.................. ");
    let test5_result: Result<bool> = (|| async {
        let test_agent = TestAgent::new();
        let agent_clone = Arc::clone(&test_agent.agent);
        
        // Start the agent's gRPC server in the background
        tokio::spawn(async move {
            if let Err(e) = agent_clone.serve(50057).await {
                eprintln!("Failed to serve: {}", e);
            }
        });
        
        // Give it a moment to start and register
        sleep(Duration::from_secs(2)).await;
        
        // If we got here without crashing, it's working
        Ok(true)
    })().await;
    
    match test5_result {
        Ok(passed) => {
            results.insert("gRPC Server", passed);
            if passed {
                println!("PASS");
                println!("   Agent gRPC server started on port 50057 and registered with control plane");
            } else {
                println!("FAIL");
            }
        }
        Err(e) => {
            results.insert("gRPC Server", false);
            println!("FAIL ({})", e);
        }
    }

    // Summary
    let passed = results.values().filter(|&&v| v).count();
    let total = results.len();
    println!("\nSummary: {}/{} tests passed", passed, total);

    Ok(passed == total)
}

#[tokio::main]
async fn main() {
    match run_standardized_tests().await {
        Ok(success) => {
            if !success {
                std::process::exit(1);
            }
        }
        Err(e) => {
            eprintln!("Test suite failed: {}", e);
            std::process::exit(1);
        }
    }
}